Reworking the `elements` package

barklund opened the pull request:
## Context

More work on the elements package, as well as deprecating the types package completely.

(this is slightly WIP - the git merge was a bit muddy to get an overview of, so might make additional changes after creating this PR)

## User-facing changes

_None (I hope!)_

## Testing Instructions

- Full (single-browser) smoke-test is probably a good idea, as this touches on a lot of things through the entire application.

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12126

Commit:
Moved types around and converted elements package

Commit:
WIP: convert element-library package

Commit:
Merge branch 'main' into code/12126-ts-elements

# Conflicts:
#	package-lock.json
#	packages/elements/src/elementType.js
#	packages/elements/src/elementType.ts
#	packages/elements/src/types.js
#	packages/elements/src/types.ts
#	packages/elements/src/types/element.ts
#	packages/elements/src/types/page.ts
#	packages/elements/src/types/propTypes.ts
#	packages/elements/src/utils/createNewElement.ts
#	packages/elements/src/utils/createPage.ts
#	packages/elements/src/utils/duplicateElement.ts
#	packages/elements/src/utils/duplicatePage.ts
#	packages/elements/src/utils/getElementOffsets.ts
#	packages/elements/src/utils/getElementOrigin.ts
#	packages/elements/src/utils/getLayerName.ts
#	packages/elements/src/utils/getTransformFlip.ts
#	packages/elements/src/utils/isElementBelowLimit.ts
#	packages/elements/tsconfig.json
#	packages/migration/package.json
#	packages/templates/src/raw/a-day-in-the-life/index.ts
#	packages/units/src/context.ts
#	packages/units/src/dimensions.ts
#	packages/units/src/getBoundRect.ts
#	packages/units/src/types.ts
#	tsconfig.json

Commit:
Re-moved types to relevant packages and untyped the element-library

Commit:
Updating package.json

swissspidy commented on packages/elements/src/utils/getElementOrigin.ts:
```diff
@@ -23,6 +23,7 @@
  * @param offsets Media element offsets
  * @return object containing horizontal and vertical transform origin percentages
  */
+
```
Errant space; now the docblock is no longer actually attached to the function.

swissspidy commented on packages/elements/tsconfig.json:
```diff
@@ -7,8 +7,7 @@
     { "path": "../i18n" },
     { "path": "../media" },
-    { "path": "../patterns" },
-    { "path": "../units" }
```
Since the `units` package is still listed as a dependency in `package.json`, we should keep the reference here as well.

barklund replied:
Yes, that's a mistake.

swissspidy commented on packages/elements/src/utils/isElementBelowLimit.ts:
```diff
@@ -33,10 +35,6 @@ function isElementBelowLimit(element: Element, verifyLink = true) {
-      return (points[point] as typeof points[keyof typeof points]).y > limit;
-    }
-  );
+  return Boolean(Object.values(points).find((coord) => coord.y > limit));
```
Thanks for simplifying this one 😄 

But why `Boolean( .find() )` when `.some()` exists?

barklund replied:
Good question, I'll simplify further.

swissspidy commented on packages/media/src/types/resource.ts:
```diff
@@ -49,7 +49,7 @@ export interface ResourceSize {
   height: number;
 }
 
-export type ResourceId = string | number;
```
This was `string | number` for a reason.

Media3p resources typically have string IDs, whereas WordPress media resources have numeric values.

barklund replied:
Sorry, a lot of the diff from here on down are due to the merge conflicts and not actually because I wanted to change what you had done, I just kept what I had. I'll go over everything here in more detail today.

swissspidy commented on packages/media/src/types/resource.ts:
```diff
@@ -77,20 +77,13 @@ export interface Resource {
-  isPlaceholder?: boolean;
+  isPlaceholder: boolean;
   /** Whether the resource needs a CORS proxy. */
-  needsProxy?: boolean;
```
IIRC these were made optional because especially in templates not all resources have this property. I think we only added in later and didn't add a data migration because it was kinda unnecessary.

So to reflect that, we should keep them optional, or add a migration.

swissspidy commented on packages/media/src/types/resource.ts:
```diff
@@ -77,20 +77,13 @@ export interface Resource {
   /** Resource creation date. */
   readonly creationDate?: string;
   /** Resource sizes */
-  sizes: { [key: string]: ResourceSize } | never[];
```
This `never[]` was added because of the templates, where images don't have multiple sizes and instead have `"sizes": [],` (check out the template JSON files to see it)

If we change this, we should add a migration & some code to turn this into `sizes: {}` if there are no sizes or making it optional.

swissspidy commented on packages/media/src/types/resource.ts:
```diff
@@ -77,20 +77,13 @@ export interface Resource {
-
-  // TODO: Figure out why sometimes _images_ end up having these properties.
-  posterId?: ResourceId;
-  isOptimized?: boolean;
```
Also added because of the templates IIRC.

If we remove this here, we should also remove these properties from the affected images in the templates.

swissspidy commented on packages/media/src/types/resource.ts:
```diff
@@ -77,20 +77,13 @@ export interface Resource {
   /** Resource author attribution */
   attribution?: Attribution;
-
-  // TODO: This property should probably not end up in a story. Only relevant for the media library.
-  provider?: 'local' | 'unsplash' | 'coverr' | 'tenor' | 'tenor_stickers';
```
If we remove this here, we should look into fixing this, hence the TODO.

swissspidy commented on packages/media/src/types/resourceInput.ts:
```diff
@@ -0,0 +1,75 @@
+   * The resource ID.
+   * TODO: currently this value is local to the editor's media system.
+   */
+  id: string;
```
Can be a number as well (hence the `ResourceId` type)

swissspidy reviewed (commented):
An additional note:

We have separate `elements` and `element-library` packages for a reason, so that these are decoupled.

We should not put type definitions for individual element types back into `elements` package when we have this dedicated package.

Commit:
Merge branch 'main' into code/12126-ts-elements

Commit:
Feedback improvements

Commit:
Fixed more from PR feedback

Commit:
Merge branch 'main' into code/12126-ts-elements

Commit:
Fix lint error

Commit:
Merge branch 'main' into code/12126-ts-elements

Commit:
Moved element definitions to element library package and updated dependencies

swissspidy commented on packages/elements/src/types/element.ts:
```diff
@@ -40,6 +46,8 @@ export interface Mask {
+  iconPath?: string;
+  iconRatio?: number;
   supportsBorder?: boolean;
 }
```
Wondering: should the `Mask` type better live in the `masks` package?

barklund replied:
The `masks` package depends on the `elements` package, and that dependency is pretty locked-in. There are all the components for display, frame, and output, that have element typings as well as several util functions (`shouldDisplayBorder`, `canMaskHaveBorder`, ...). And the big one is the `BorderedMaskedElement` using many element features.

Some of these could be solved by utilizing `DimensionableElement` from units package (though the border in that package does not include a color which masks need, which would add a dependency on the patterns package, but that's probably okay), others could be solved by also adding a `MaskableElement` to the masks package, but we still also need things like the `flip` and `isBackground` properties from the elements (in the `BorderedMaskedElement`), so that would require making even more element interfaces.

It's doable, but it's getting a bit complex (a naming is getting terrible 😂). Let's move this idea to a new ticket.

swissspidy commented on packages/media/src/types/sequenceResource.ts:
```diff
@@ -17,14 +17,14 @@
   poster?: string;
   /** The resource's poster ID. */
-  posterId?: ResourceId;
+  posterId?: string;
```
This should continue to use `ResourceId`.

barklund replied:
Fixed in 1e3577e885.

swissspidy commented on packages/elements/src/types/elementDefinition.ts:
```diff
@@ -0,0 +1,96 @@
+  element: E;
+}
+
+export interface ElementDefinition<E extends Element = Element> {
```
I like the refactoring here with generics! 👍 

Looking forward to using these types and see how well it works.

swissspidy reviewed (commented):
There are some test failures & prop types warnings, but otherwise looks good!

miina commented on packages/elements/src/types/elementDefinition.ts:
```diff
@@ -0,0 +1,96 @@
+  isMedia?: boolean;
+  getLayerText: (element: E) => string;
+  defaultAttributes: Partial<E>;
+  Edit: React.FC<EditProps<E>>;
```
I've seen in multiple articles / resources mentioned that `React.FC` is better not to use, e.g. because it causes the component to implicitly accept children. I _think_ there were some other reasons as well. Thoughts? You probably have more information about it.

barklund replied:
That has been [changed in React 18](https://solverfox.dev/writing/no-implicit-children/). `React.FC` (or `React.FunctionComponent`) no longer implicitly accepts children.

**BUT** we're _not_ using React 18, and probably never will because of all the migration issues. So you're right, it should probably be changed to a more clean type.

barklund replied:
`React.VoidFunctionComponent` seems to be [the way](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34237#issuecomment-915178375) to do it pre-18. Thanks!

barklund replied:
Fixed in 1e3577e885.

miina replied:
Nice! Should we add the eslint configuration as well to prefer this to `React.FC` mentioned in the [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34237#issuecomment-915178375) you shared? cc @swissspidy

swissspidy replied:
Sure, why not, perhaps in a new PR?

Commit:
Fixed React-17 React.FC issue and resource id reference

miina reviewed (approved):
Nice rework! The JS tests are failing, otherwise LGTM 🐘

Commit:
Removed Page.type property as it is redundant

Also fixed a test.

Note that there are many more instance of type:"page" in the code, but most are related to tests that will probably still work fine or templates, which will also work fine with an extra unused property. Should we add a migration though?

Commit:
Fix imports of background audio prop type

Commit:
Fixed a karma test

Commit:
Finally fixed that annoying bg-shape-element copy thingy

Commit:
Added a comment to make sense of it

merapi committed:
Fix eyedropper

merapi committed:
Target correct element

Commit:
Merge branch 'main' into code/12126-ts-elements

# Conflicts:
#	tsconfig.json

Commit:
Merge remote-tracking branch 'origin/code/12126-ts-elements' into code/12126-ts-elements

Commit:
Updated type references for config providers

Commit:
Missing types in package.json?

Commit:
Merge branch 'main' into code/12126-ts-elements

# Conflicts:
#	packages/media/src/createResource.ts
#	packages/media/src/types/propTypes.ts
#	packages/wp-story-editor/src/api/utils/normalizeResourceSizes.js

Commit:
Reorganized data types to elements package
